internal/component/loki/source/file/file.go

```diff
@@ -151,30 +185,53 @@ func (c *Component) Run(ctx context.Context) error {
 				}
 			}
 		}()
-		c.schedulerMut.Lock()
+		c.mut.Lock()
 		c.stopping.Store(true)
+		c.watcher.Stop()
 		c.scheduler.Stop()
 		close(c.handler.Chan())
-		c.schedulerMut.Unlock()
+		c.mut.Unlock()
 	}()
 
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case entry := <-c.handler.Chan():
-			c.receiversMut.RLock()
-			for _, receiver := range c.receivers {
-				select {
-				case <-ctx.Done():
-					c.receiversMut.RUnlock()
-					return nil
-				case receiver.Chan() <- entry:
+	var wg sync.WaitGroup
+	wg.Go(func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case entry := <-c.handler.Chan():
+				c.receiversMut.RLock()
+				for _, receiver := range c.receivers {
+					select {
+					case <-ctx.Done():
+						c.receiversMut.RUnlock()
+						return
+					case receiver.Chan() <- entry:
+					}
 				}
+				c.receiversMut.RUnlock()
 			}
-			c.receiversMut.RUnlock()
 		}
-	}
+	})
+
+	wg.Go(func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-c.watcher.C:
+				c.mut.Lock()
+				if !c.args.FileMatch.Enabled {
+					return
+				}
```

Comment:
So if we disable FileMatch and then enable FileMatch, I think the goroutine will not be restarted... can we add a test for this and fix the issue? 

We have two choices: always keep the goroutines and watcher ticker on and check for FileMatch enabled - a bit more expensive but a simpler code. Or we schedule them only when FileMatch is enabled - a bit more complex, but fewer resources used. I don't have a strong opinion, I would like this code to be easier to follow though...